import sys
import math as ms
from PyQt5 import QtCore, QtGui, QtWidgets


class Ui_Calculator(object):
    def setupUi(self, Calculator):
        Calculator.setObjectName("Calculator")
        Calculator.setFixedSize(444, 439)
        Calculator.setStyleSheet("background-color: rgb(0, 0, 0);")
        self.centralwidget = QtWidgets.QWidget(Calculator)
        self.centralwidget.setObjectName("centralwidget")
        self.log2 = QtWidgets.QPushButton(self.centralwidget)
        self.log2.setGeometry(QtCore.QRect(20, 220, 61, 61))
        font = QtGui.QFont()
        font.setFamily("Sitka Small")
        font.setPointSize(22)
        font.setBold(False)
        font.setWeight(50)
        self.log2.setFont(font)
        self.log2.setStyleSheet("border-radius:10px;\n"
"background-color: rgb(188, 188, 188);")
        self.log2.setObjectName("log2")
        self.one = QtWidgets.QPushButton(self.centralwidget)
        self.one.setGeometry(QtCore.QRect(90, 150, 61, 61))
        font = QtGui.QFont()
        font.setFamily("Sitka Small")
        font.setPointSize(24)
        font.setBold(False)
        font.setWeight(50)
        self.one.setFont(font)
        self.one.setStyleSheet("border-radius:10px;\n"
"background-color: rgb(108, 108, 108);")
        self.one.setObjectName("one")
        self.plus = QtWidgets.QPushButton(self.centralwidget)
        self.plus.setGeometry(QtCore.QRect(300, 220, 61, 61))
        font = QtGui.QFont()
        font.setFamily("Sitka Small")
        font.setPointSize(24)
        font.setBold(False)
        font.setWeight(50)
        self.plus.setFont(font)
        self.plus.setStyleSheet("border-radius:10px;\n"
"background-color: rgb(188, 188, 188);")
        self.plus.setObjectName("plus")
        self.divide = QtWidgets.QPushButton(self.centralwidget)
        self.divide.setGeometry(QtCore.QRect(300, 290, 61, 61))
        font = QtGui.QFont()
        font.setFamily("Sitka Small")
        font.setPointSize(24)
        font.setBold(False)
        font.setWeight(50)
        self.divide.setFont(font)
        self.divide.setStyleSheet("border-radius:10px;\n"
"background-color: rgb(188, 188, 188);")
        self.divide.setObjectName("divide")
        self.multiply = QtWidgets.QPushButton(self.centralwidget)
        self.multiply.setGeometry(QtCore.QRect(300, 360, 61, 61))
        font = QtGui.QFont()
        font.setFamily("Sitka Small")
        font.setPointSize(24)
        font.setBold(False)
        font.setWeight(50)
        self.multiply.setFont(font)
        self.multiply.setStyleSheet("border-radius:10px;\n"
"background-color: rgb(188, 188, 188);")
        self.multiply.setObjectName("multiply")
        self.factorial = QtWidgets.QPushButton(self.centralwidget)
        self.factorial.setGeometry(QtCore.QRect(230, 360, 61, 61))
        font = QtGui.QFont()
        font.setFamily("Sitka Small")
        font.setPointSize(24)
        font.setBold(False)
        font.setWeight(50)
        self.factorial.setFont(font)
        self.factorial.setStyleSheet("border-radius:10px;\n"
"background-color: rgb(188, 188, 188);")
        self.factorial.setObjectName("factorial")
        self.root = QtWidgets.QPushButton(self.centralwidget)
        self.root.setGeometry(QtCore.QRect(90, 360, 61, 61))
        font = QtGui.QFont()
        font.setFamily("Sitka Small")
        font.setPointSize(24)
        font.setBold(False)
        font.setWeight(50)
        self.root.setFont(font)
        self.root.setStyleSheet("border-radius:10px;\n"
"background-color: rgb(188, 188, 188);")
        self.root.setObjectName("root")
        self.lg10 = QtWidgets.QPushButton(self.centralwidget)
        self.lg10.setGeometry(QtCore.QRect(20, 290, 61, 61))
        font = QtGui.QFont()
        font.setFamily("Sitka Small")
        font.setPointSize(22)
        font.setBold(False)
        font.setWeight(50)
        self.lg10.setFont(font)
        self.lg10.setStyleSheet("border-radius:10px;\n"
"background-color: rgb(188, 188, 188);")
        self.lg10.setObjectName("lg10")
        self.fabs = QtWidgets.QPushButton(self.centralwidget)
        self.fabs.setGeometry(QtCore.QRect(20, 360, 61, 61))
        font = QtGui.QFont()
        font.setFamily("Sitka Small")
        font.setPointSize(22)
        font.setBold(False)
        font.setWeight(50)
        self.fabs.setFont(font)
        self.fabs.setStyleSheet("border-radius:10px;\n"
"background-color: rgb(188, 188, 188);")
        self.fabs.setObjectName("fabs")
        self.minus = QtWidgets.QPushButton(self.centralwidget)
        self.minus.setGeometry(QtCore.QRect(300, 150, 61, 61))
        font = QtGui.QFont()
        font.setFamily("Sitka Small")
        font.setPointSize(24)
        font.setBold(False)
        font.setWeight(50)
        self.minus.setFont(font)
        self.minus.setStyleSheet("border-radius:10px;\n"
"background-color: rgb(188, 188, 188);")
        self.minus.setObjectName("minus")
        self.log = QtWidgets.QPushButton(self.centralwidget)
        self.log.setGeometry(QtCore.QRect(20, 150, 61, 61))
        font = QtGui.QFont()
        font.setFamily("Sitka Small")
        font.setPointSize(24)
        font.setBold(False)
        font.setWeight(50)
        self.log.setFont(font)
        self.log.setStyleSheet("border-radius:10px;\n"
"background-color: rgb(188, 188, 188);")
        self.log.setObjectName("log")
        self.two = QtWidgets.QPushButton(self.centralwidget)
        self.two.setGeometry(QtCore.QRect(160, 150, 61, 61))
        font = QtGui.QFont()
        font.setFamily("Sitka Small")
        font.setPointSize(24)
        font.setBold(False)
        font.setWeight(50)
        self.two.setFont(font)
        self.two.setStyleSheet("border-radius:10px;\n"
"background-color: rgb(108, 108, 108);")
        self.two.setObjectName("two")
        self.three = QtWidgets.QPushButton(self.centralwidget)
        self.three.setGeometry(QtCore.QRect(230, 150, 61, 61))
        font = QtGui.QFont()
        font.setFamily("Sitka Small")
        font.setPointSize(24)
        font.setBold(False)
        font.setWeight(50)
        self.three.setFont(font)
        self.three.setStyleSheet("border-radius:10px;\n"
"background-color: rgb(108, 108, 108);")
        self.three.setObjectName("three")
        self.six = QtWidgets.QPushButton(self.centralwidget)
        self.six.setGeometry(QtCore.QRect(230, 220, 61, 61))
        font = QtGui.QFont()
        font.setFamily("Sitka Small")
        font.setPointSize(24)
        font.setBold(False)
        font.setWeight(50)
        self.six.setFont(font)
        self.six.setStyleSheet("border-radius:10px;\n"
"background-color: rgb(108, 108, 108);")
        self.six.setObjectName("six")
        self.seven = QtWidgets.QPushButton(self.centralwidget)
        self.seven.setGeometry(QtCore.QRect(90, 290, 61, 61))
        font = QtGui.QFont()
        font.setFamily("Sitka Small")
        font.setPointSize(24)
        font.setBold(False)
        font.setWeight(50)
        self.seven.setFont(font)
        self.seven.setStyleSheet("border-radius:10px;\n"
"background-color: rgb(108, 108, 108);")
        self.seven.setObjectName("seven")
        self.zero = QtWidgets.QPushButton(self.centralwidget)
        self.zero.setGeometry(QtCore.QRect(160, 360, 61, 61))
        font = QtGui.QFont()
        font.setFamily("Sitka Small")
        font.setPointSize(24)
        font.setBold(False)
        font.setWeight(50)
        self.zero.setFont(font)
        self.zero.setStyleSheet("border-radius:10px;\n"
"background-color: rgb(108, 108, 108);")
        self.zero.setObjectName("zero")
        self.eight = QtWidgets.QPushButton(self.centralwidget)
        self.eight.setGeometry(QtCore.QRect(160, 290, 61, 61))
        font = QtGui.QFont()
        font.setFamily("Sitka Small")
        font.setPointSize(24)
        font.setBold(False)
        font.setWeight(50)
        self.eight.setFont(font)
        self.eight.setStyleSheet("border-radius:10px;\n"
"background-color: rgb(108, 108, 108);")
        self.eight.setObjectName("eight")
        self.nine = QtWidgets.QPushButton(self.centralwidget)
        self.nine.setGeometry(QtCore.QRect(230, 290, 61, 61))
        font = QtGui.QFont()
        font.setFamily("Sitka Small")
        font.setPointSize(24)
        font.setBold(False)
        font.setWeight(50)
        self.nine.setFont(font)
        self.nine.setStyleSheet("border-radius:10px;\n"
"background-color: rgb(108, 108, 108);")
        self.nine.setObjectName("nine")
        self.four = QtWidgets.QPushButton(self.centralwidget)
        self.four.setGeometry(QtCore.QRect(90, 220, 61, 61))
        font = QtGui.QFont()
        font.setFamily("Sitka Small")
        font.setPointSize(24)
        font.setBold(False)
        font.setWeight(50)
        self.four.setFont(font)
        self.four.setStyleSheet("border-radius:10px;\n"
"background-color: rgb(108, 108, 108);")
        self.four.setObjectName("four")
        self.five = QtWidgets.QPushButton(self.centralwidget)
        self.five.setGeometry(QtCore.QRect(160, 220, 61, 61))
        font = QtGui.QFont()
        font.setFamily("Sitka Small")
        font.setPointSize(24)
        font.setBold(False)
        font.setWeight(50)
        self.five.setFont(font)
        self.five.setStyleSheet("border-radius:10px;\n"
"background-color: rgb(108, 108, 108);")
        self.five.setObjectName("five")
        self.otvet = QtWidgets.QLabel(self.centralwidget)
        self.otvet.setGeometry(QtCore.QRect(40, 80, 371, 51))
        font = QtGui.QFont()
        font.setFamily("Sitka Small")
        font.setPointSize(26)
        self.otvet.setFont(font)
        self.otvet.setStyleSheet("color: rgb(255, 255, 255);")
        self.otvet.setObjectName("otvet")
        self.prd_otvet = QtWidgets.QLabel(self.centralwidget)
        self.prd_otvet.setGeometry(QtCore.QRect(40, 20, 371, 51))
        font = QtGui.QFont()
        font.setFamily("Sitka Small")
        font.setPointSize(26)
        self.prd_otvet.setFont(font)
        self.prd_otvet.setStyleSheet("color: rgb(185, 185, 185);")
        self.prd_otvet.setObjectName("prd_otvet")
        self.clear = QtWidgets.QPushButton(self.centralwidget)
        self.clear.setGeometry(QtCore.QRect(370, 150, 61, 61))
        font = QtGui.QFont()
        font.setFamily("Sitka Small")
        font.setPointSize(24)
        font.setBold(False)
        font.setWeight(50)
        self.clear.setFont(font)
        self.clear.setStyleSheet("border-radius:10px;\n"
"background-color: rgb(217, 145, 0);")
        self.clear.setObjectName("clear")
        self.knop_men = QtWidgets.QPushButton(self.centralwidget)
        self.knop_men.setGeometry(QtCore.QRect(370, 220, 61, 61))
        font = QtGui.QFont()
        font.setFamily("Sitka Small")
        font.setPointSize(24)
        font.setBold(False)
        font.setWeight(50)
        self.knop_men.setFont(font)
        self.knop_men.setStyleSheet("border-radius:10px;\n"
"background-color: rgb(188, 188, 188);")
        self.knop_men.setObjectName("knop_men")
        self.PI = QtWidgets.QPushButton(self.centralwidget)
        self.PI.setGeometry(QtCore.QRect(370, 290, 61, 61))
        font = QtGui.QFont()
        font.setFamily("Sitka Small")
        font.setPointSize(24)
        font.setBold(False)
        font.setWeight(50)
        self.PI.setFont(font)
        self.PI.setStyleSheet("border-radius:10px;\n"
"background-color: rgb(188, 188, 188);")
        self.PI.setObjectName("PI")
        self.ravno = QtWidgets.QPushButton(self.centralwidget)
        self.ravno.setGeometry(QtCore.QRect(370, 360, 61, 61))
        font = QtGui.QFont()
        font.setFamily("Sitka Small")
        font.setPointSize(24)
        font.setBold(False)
        font.setWeight(50)
        self.ravno.setFont(font)
        self.ravno.setStyleSheet("border-radius:10px;\n"
"background-color: rgb(188, 188, 188);")
        self.ravno.setObjectName("ravno")
        Calculator.setCentralWidget(self.centralwidget)

        self.retranslateUi(Calculator)
        QtCore.QMetaObject.connectSlotsByName(Calculator)

    def retranslateUi(self, Calculator):
        _translate = QtCore.QCoreApplication.translate
        Calculator.setWindowTitle(_translate("Calculator", "Home calculator"))
        self.log2.setText(_translate("Calculator", "log2"))
        self.one.setText(_translate("Calculator", "1"))
        self.plus.setText(_translate("Calculator", "+"))
        self.divide.setText(_translate("Calculator", "/"))
        self.multiply.setText(_translate("Calculator", "x"))
        self.factorial.setText(_translate("Calculator", "n!"))
        self.root.setText(_translate("Calculator", "√"))
        self.lg10.setText(_translate("Calculator", "lg10"))
        self.fabs.setText(_translate("Calculator", "[x]"))
        self.minus.setText(_translate("Calculator", "-"))
        self.log.setText(_translate("Calculator", "log"))
        self.two.setText(_translate("Calculator", "2"))
        self.three.setText(_translate("Calculator", "3"))
        self.six.setText(_translate("Calculator", "6"))
        self.seven.setText(_translate("Calculator", "7"))
        self.zero.setText(_translate("Calculator", "0"))
        self.eight.setText(_translate("Calculator", "8"))
        self.nine.setText(_translate("Calculator", "9"))
        self.four.setText(_translate("Calculator", "4"))
        self.five.setText(_translate("Calculator", "5"))
        self.otvet.setText(_translate("Calculator", ""))
        self.prd_otvet.setText(_translate("Calculator", ""))
        self.clear.setText(_translate("Calculator", "AC"))
        self.knop_men.setText(_translate("Calculator", "<"))
        self.PI.setText(_translate("Calculator", "π"))
        self.ravno.setText(_translate("Calculator", "="))

class Window(QtWidgets.QMainWindow):  #подключение к кнопкам
    def __init__(self):
        super(Window,self).__init__()
        self.ui = Ui_Calculator()
        self.ui.setupUi(self)
        self.ui.one.clicked.connect(lambda: self.Numbut(1))
        self.ui.two.clicked.connect(lambda: self.Numbut(2))
        self.ui.three.clicked.connect(lambda: self.Numbut(3))
        self.ui.four.clicked.connect(lambda: self.Numbut(4))
        self.ui.five.clicked.connect(lambda: self.Numbut(5))
        self.ui.six.clicked.connect(lambda: self.Numbut(6))
        self.ui.seven.clicked.connect(lambda: self.Numbut(7))
        self.ui.eight.clicked.connect(lambda: self.Numbut(8))
        self.ui.nine.clicked.connect(lambda: self.Numbut(9))
        self.ui.zero.clicked.connect(lambda: self.Numbut(0))
        self.ui.PI.clicked.connect(lambda: self.Numbut(3.14))
        self.ui.minus.clicked.connect(lambda: self.Numbut("-"))
        self.ui.plus.clicked.connect(lambda: self.Numbut("+"))
        self.ui.divide.clicked.connect(lambda: self.Numbut("/"))
        self.ui.multiply.clicked.connect(lambda: self.Numbut("*"))
        self.ui.factorial.clicked.connect(self.Factorial)
        self.ui.clear.clicked.connect(self.Clear)
        self.ui.ravno.clicked.connect(self.Ravno)
        self.ui.log.clicked.connect(self.Log)
        self.ui.log2.clicked.connect(self.Log2)
        self.ui.lg10.clicked.connect(self.Lg10)
        self.ui.fabs.clicked.connect(self.Fabs)
        self.ui.root.clicked.connect(self.Coren)
        self.ui.knop_men.clicked.connect(self.Delete)



    def Ravno(self):                                    #функция вычисления и нажатия на кнопку равно для переведения последнего ответа в другое текстовое поле
        minus = "-"
        plus = "+"
        multiply = "*"
        devide = "/"
        Compare = self.ui.otvet.text()
        if minus in Compare:
                s = Compare.split("-")
                n = float(s[0])
                n1 = float(s[1])
                f = float(n - n1)
                self.ui.otvet.setText(str(f))
        elif plus in Compare:
                s = Compare.split("+")
                n = float(s[0])
                n1 = float(s[1])
                f = float(n + n1)
                self.ui.otvet.setText(str(f))
        elif devide in Compare:
                s = Compare.split("/")
                n = float(s[0])
                n1 = float(s[1])
                f = float(n / n1)
                self.ui.otvet.setText(str(f))
        elif multiply in Compare:
                s = Compare.split("*")
                n = float(s[0])
                n1 = float(s[1])
                f = float(n * n1)
                self.ui.otvet.setText(str(f))

        self.ui.prd_otvet.setText(self.ui.otvet.text())
        self.ui.otvet.clear()

    def Clear(self):                                      #функция отчистки поля
        self.ui.otvet.clear()

    def Factorial(self):                                  #функция факториала
        N = ms.factorial(int(self.ui.otvet.text()))
        print (N)
        self.ui.otvet.setText(str(N))

    def Log(self):                                        #функция логорифма
        N = ms.log(int(self.ui.otvet.text()))
        print(N)
        self.ui.otvet.setText(str(N))

    def Log2(self):                                       #функция логорифма2
        N = ms.log2(int(self.ui.otvet.text()))
        print(N)
        self.ui.otvet.setText(str(N))

    def Lg10(self):                                       #функция логорифма10
        N = ms.log10(int(self.ui.otvet.text()))
        print(N)
        self.ui.otvet.setText(str(N))

    def Fabs(self):                                       #функция модуля числа
        N = ms.fabs(int(self.ui.otvet.text()))
        print(N)
        self.ui.otvet.setText(str(N))

    def Coren(self):                                      #функция корня
        N = ms.sqrt(int(self.ui.otvet.text()))
        print(N)
        self.ui.otvet.setText(str(N))

    def Numbut(self, num):                                #функция ввода чисел в текстовую строку
        v = str(num)
        self.ui.otvet.setText(self.ui.otvet.text()+v)

    def Delete(self):                                     #функция удаления символов
        a = self.ui.otvet.text()
        a= a[:-1]
        self.ui.otvet.setText(a)


app = QtWidgets.QApplication([])
application = Window()
application.show()
sys.exit(app.exec())